thunder/executors/torch_autograd.py

```diff
@@ -19,8 +19,11 @@
 
 class ThunderFunction(torch.autograd.Function):
     @staticmethod
-    def forward(ctx, compiled_backward, saved_tensors, saved_other, flat_output, *flat_args):
+    def forward(
+        ctx, return_none_instead_of_grads, compiled_backward, saved_tensors, saved_other, flat_output, *flat_args
+    ):
         # Here we just propagate the tensors through the autograd graph
+        ctx.return_none_instead_of_grads = return_none_instead_of_grads
```

Comment:
How about naming this `is_fsdp_no_sync_enabled`? That way it will be easier to understand what this flag means outside of `ThunderFunction` code.

Reply:
An alternative solution to returning Nones here would be modifying parameters in place to make them unsharded while inside the no_sync context and shard them again when exiting the context.
What do you think, would this work? We can access parameter tensors now within impl functions as done in `_stash_grad_for_fsdp_prim_impl`.

Reply:
> modifying parameters in place to make them unsharded while inside the no_sync context and shard them again when exiting the context

iiuc it would further bloat up the memory footprint thus I'm not inclined to it.

Reply:
For ZeRO2 I think it would even be less memory-consuming because today we hold both sharded and unsharded parameters after forward.
For ZeRO3 we could do this inplace unsharding only during backward.